File: internal/vault/frontmatter.go
// Package vault provides core Obsidian vault operations.
package vault

import (
	"strings"

	"gopkg.in/yaml.v3"
)

// ParsedNote holds a note split into its frontmatter, content, and raw text.
type ParsedNote struct {
	Frontmatter map[string]interface{} // parsed YAML metadata
	Content     string                 // everything after the frontmatter
	Raw         string                 // original full text
}

// ParseNote splits a markdown note into frontmatter and content.
// Frontmatter is YAML between --- delimiters at the start.
func ParseNote(raw string) ParsedNote {
	lines := strings.Split(raw, "\n")

	if len(lines) == 0 || strings.TrimSpace(lines[0]) != "---" {
		return ParsedNote{Frontmatter: map[string]interface{}{}, Content: raw, Raw: raw}
	}

	// Find closing ---
	closeIndex := -1
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(lines[i]) == "---" {
			closeIndex = i
			break
		}
	}

	if closeIndex == -1 {
		return ParsedNote{Frontmatter: map[string]interface{}{}, Content: raw, Raw: raw}
	}

	// Parse YAML frontmatter
	fmText := strings.Join(lines[1:closeIndex], "\n")
	frontmatter := map[string]interface{}{}
	if strings.TrimSpace(fmText) != "" {
		if err := yaml.Unmarshal([]byte(fmText), &frontmatter); err != nil {
			frontmatter = map[string]interface{}{}
		}
	}

	// Content is everything after closing ---
	content := ""
	if closeIndex+1 < len(lines) {
		content = strings.TrimLeft(strings.Join(lines[closeIndex+1:], "\n"), "\n")
	}

	return ParsedNote{Frontmatter: frontmatter, Content: content, Raw: raw}
}

// SerializeNote reconstructs markdown from frontmatter and content.
func SerializeNote(frontmatter map[string]interface{}, content string) string {
	if len(frontmatter) == 0 {
		return content
	}

	data, err := yaml.Marshal(frontmatter)
	if err != nil {
		return content
	}

	return "---\n" + string(data) + "---\n" + content
}

// MergeTags deduplicates and merges two tag slices, preserving order.
func MergeTags(existing, toAdd []string) []string {
	seen := map[string]bool{}
	result := []string{}

	for _, t := range existing {
		if !seen[t] {
			seen[t] = true
			result = append(result, t)
		}
	}
	for _, t := range toAdd {
		if !seen[t] {
			seen[t] = true
			result = append(result, t)
		}
	}
	return result
}

// AddFrontmatterTags adds tags to a note's frontmatter, merging without duplicates.
func AddFrontmatterTags(note ParsedNote, tags []string) ParsedNote {
	current := note.Frontmatter["tags"]
	var currentSlice []string

	switch v := current.(type) {
	case []interface{}:
		for _, item := range v {
			if s, ok := item.(string); ok {
				currentSlice = append(currentSlice, s)
			}
		}
	case []string:
		currentSlice = v
	case string:
		if v != "" {
			currentSlice = []string{v}
		}
	}

	merged := MergeTags(currentSlice, tags)

	newFM := make(map[string]interface{}, len(note.Frontmatter))
	for k, v := range note.Frontmatter {
		newFM[k] = v
	}
	newFM["tags"] = merged

	return ParsedNote{
		Frontmatter: newFM,
		Content:     note.Content,
		Raw:         note.Raw,
	}
}

// ExtractWikilinks returns all [[link]] targets from text.
func ExtractWikilinks(content string) []string {
	var links []string
	remaining := content
	for {
		start := strings.Index(remaining, "[[")
		if start == -1 {
			break
		}
		end := strings.Index(remaining[start:], "]]")
		if end == -1 {
			break
		}
		inner := remaining[start+2 : start+end]
		// Handle [[link|display]] — take only the link part
		if pipe := strings.Index(inner, "|"); pipe != -1 {
			inner = inner[:pipe]
		}
		links = append(links, strings.TrimSpace(inner))
		remaining = remaining[start+end+2:]
	}
	return links
}

// GetTags returns the tags from a parsed note's frontmatter as a []string.
func GetTags(note ParsedNote) []string {
	current := note.Frontmatter["tags"]
	switch v := current.(type) {
	case []interface{}:
		tags := make([]string, 0, len(v))
		for _, item := range v {
			if s, ok := item.(string); ok {
				tags = append(tags, s)
			}
		}
		return tags
	case []string:
		return v
	case string:
		if v != "" {
			return []string{v}
		}
	}
	return nil
}

File: src/vault/frontmatter.ts
import { parse as parseYaml, stringify as stringifyYaml } from 'yaml';

export interface ParsedNote {
  frontmatter: Record<string, any>;  // parsed YAML
  content: string;                     // everything after frontmatter
  raw: string;                         // original full text
}

/**
 * Parse a markdown note into frontmatter, content, and raw.
 * Frontmatter is YAML between --- delimiters at the start.
 */
export function parseNote(raw: string): ParsedNote {
  const lines = raw.split('\n');

  // Check if first line is ---
  if (lines.length === 0 || lines[0].trim() !== '---') {
    return {
      frontmatter: {},
      content: raw,
      raw,
    };
  }

  // Find closing --- delimiter
  let closeIndex = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === '---') {
      closeIndex = i;
      break;
    }
  }

  if (closeIndex === -1) {
    // No closing delimiter, treat entire content as raw
    return {
      frontmatter: {},
      content: raw,
      raw,
    };
  }

  // Extract frontmatter YAML
  const frontmatterLines = lines.slice(1, closeIndex);
  const frontmatterText = frontmatterLines.join('\n');

  let frontmatter: Record<string, any> = {};
  if (frontmatterText.trim()) {
    try {
      const parsed = parseYaml(frontmatterText);
      if (parsed && typeof parsed === 'object') {
        frontmatter = parsed;
      }
    } catch {
      // If YAML parsing fails, treat as empty frontmatter
      frontmatter = {};
    }
  }

  // Extract content (everything after closing ---)
  const content = lines.slice(closeIndex + 1).join('\n').trimStart();

  return {
    frontmatter,
    content,
    raw,
  };
}

/**
 * Serialize frontmatter and content back to markdown.
 */
export function serializeNote(frontmatter: Record<string, any>, content: string): string {
  if (Object.keys(frontmatter).length === 0) {
    return content;
  }

  const yaml = stringifyYaml(frontmatter, { indent: 2 });
  return `---\n${yaml}---\n${content}`;
}

/**
 * Merge tags arrays, removing duplicates, preserving order.
 */
export function mergeTags(existing: string[], toAdd: string[]): string[] {
  const seen = new Set<string>();
  const result: string[] = [];

  for (const tag of existing) {
    if (!seen.has(tag)) {
      seen.add(tag);
      result.push(tag);
    }
  }

  for (const tag of toAdd) {
    if (!seen.has(tag)) {
      seen.add(tag);
      result.push(tag);
    }
  }

  return result;
}

/**
 * Add tags to a parsed note's frontmatter, handling both array and list formats.
 */
export function addFrontmatterTags(note: ParsedNote, tags: string[]): ParsedNote {
  const current = note.frontmatter.tags || [];

  // Normalize to array
  let currentArray: string[] = [];
  if (Array.isArray(current)) {
    currentArray = current;
  } else if (typeof current === 'string') {
    currentArray = [current];
  }

  const merged = mergeTags(currentArray, tags);

  return {
    ...note,
    frontmatter: {
      ...note.frontmatter,
      tags: merged,
    },
  };
}
